lib/srv/db/server.go

```diff
@@ -811,6 +811,23 @@ func (s *Server) handleConnection(ctx context.Context, clientConn net.Conn) erro
 
 	err = engine.HandleConnection(ctx, sessionCtx)
 	if err != nil {
+		connectionDiagnosticID := sessionCtx.Identity.ConnectionDiagnosticID
+		if connectionDiagnosticID != "" && trace.IsAccessDenied(err) {
+			_, diagErr := s.cfg.AuthClient.AppendDiagnosticTrace(ctx,
+				connectionDiagnosticID,
+				&types.ConnectionDiagnosticTrace{
+					Type:    types.ConnectionDiagnosticTrace_RBAC_DATABASE_LOGIN,
+					Status:  types.ConnectionDiagnosticTrace_FAILED,
+					Details: "Access denied when accessing Database. Ensure your role allows you to use the requested Database Name and Database User.",
```

Comment:
For cloud-hosted databases (e.g. AWS RDS), you can get `IsAccessDenied` for other reasons like failing to get generate IAM auth, or the database rejecting the IAM auth token, etc. For example:
https://github.com/gravitational/teleport/blob/c67f452fe7d58315e4af4bb813051e5ecb5e9b15/lib/srv/db/common/errors.go#L136-L145
I think those are good cases if we can catch through the diag but they might not match this description.

Reply:
Agree
I think we can add other error matchers as we add other Database Types.
For now, it seems to be correct.

Reply:
these are all Postgres though, just cloud-hosted.